File: webgl/src/js/SubSceneParticles.js
// SubSceneParticles.js

var GL             = bongiovi.GL, gl;
var ViewSave       = require("./particles/ViewSave");
var ViewRender     = require("./particles/ViewRender");
var ViewSimulation = require("./particles/ViewSimulation");

var SubSceneParticles = function(mParentScene) {
	gl                 = GL.gl;
	this.parentScene   = mParentScene;
	this.camera        = this.parentScene.camera;
	this.cameraOtho    = this.parentScene.cameraOtho;
	this.sceneRotation = this.parentScene.sceneRotation;
	this.rotationFront = this.parentScene.rotationFront;
	this._frameCount   = 0;
	this._initTextures();
	this._initViews();

	this.reset();
}


var p = SubSceneParticles.prototype;

p._initTextures = function() {
	var numParticles = params.particles.numParticles;
	var pars = { minFilter:gl.NEAREST, magFilter:gl.NEAREST };

	this._fboCurrent 	= new bongiovi.FrameBuffer(numParticles*2.0, numParticles*2.0, pars);
	this._fboTarget 	= new bongiovi.FrameBuffer(numParticles*2.0, numParticles*2.0, pars);

	GL.setViewport(0, 0, this._fboCurrent.width, this._fboCurrent.height);
	this._fboCurrent.bind();
	GL.clear(0, 0, 0, 0);
	this._fboCurrent.unbind();

	this._fboTarget.bind();
	GL.clear(0, 0, 0, 0);
	this._fboTarget.unbind();

	GL.setViewport(0, 0, GL.canvas.width, GL.canvas.height);
};


p._initViews = function() {
	this._vSave     = new ViewSave(params.particles.numParticles);
	this._vRender   = new ViewRender(params.particles.numParticles);
	this._vSim      = new ViewSimulation();
	// this._vCopy 	= new bongiovi.ViewCopy();
};


p.reset = function() {
	GL.setMatrices(this.cameraOtho);
	GL.rotate(this.rotationFront);
	GL.setViewport(0, 0, this._fboCurrent.width, this._fboCurrent.height);
	this._fboCurrent.bind();
	this._vSave.render();
	this._fboCurrent.unbind();
	GL.setMatrices(this.camera);
	GL.rotate(this.sceneRotation.matrix);

	GL.setViewport(0, 0, GL.canvas.width, GL.canvas.height);
};


p.update = function(trees) {
	GL.setMatrices(this.cameraOtho);
	GL.rotate(this.rotationFront);
	GL.setViewport(0, 0, this._fboCurrent.width, this._fboCurrent.height);
	this._fboTarget.bind();
	GL.clear(0, 0, 0, 0);
	this._vSim.render(this._fboCurrent.getTexture(), trees );
	this._fboTarget.unbind();

	this.swap();
};


p.render = function(percent) {
	this._vRender.render(this._fboTarget.getTexture(), this._fboCurrent.getTexture(), percent);
};


p.getFbo = function() {
	return this._fboCurrent;
};

p.swap = function() {
	var tmp = this._fboTarget;
	this._fboTarget = this._fboCurrent;
	this._fboCurrent = tmp;
};

module.exports = SubSceneParticles;

File: cinder/Entrainment/src/SubSceneParticles.cpp
    //
//  SubSceneParticles.cpp
//  Entrainment03
//
//  Created by Yi-Wen Lin on 2015/7/28.
//
//

#include "SubSceneParticles.h"
#include "GlobalSettings.h"
#include "Utils.h"

using namespace bongiovi::utils;


SubSceneParticles::SubSceneParticles() {
    
}


SubSceneParticles::SubSceneParticles(CameraPersp* camera, CameraOrtho* cameraOtho, SceneQuat* sceneQuat, vector<Vec3f> treePos) : _camera(camera), _cameraOtho(cameraOtho), _sceneQuat(sceneQuat), _treePos(treePos) {
    
    _init();
}


void SubSceneParticles::_init() {
    
    Fbo::Format format;
    format.setColorInternalFormat(GL_RGBA32F_ARB);
    format.setMinFilter(GL_NEAREST);
    format.setMagFilter(GL_NEAREST);
    int size = GlobalSettings::getInstance().numParticles;
    _fboCurrent = new Fbo(size*2, size*2, format);
    _fboTarget = new Fbo(size*2, size*2, format);
    
    _fboCurrent->bindFramebuffer();
    gl::clear();
    _fboCurrent->unbindFramebuffer();
    
    _fboTarget->bindFramebuffer();
    gl::clear();
    _fboTarget->unbindFramebuffer();
    
    
    //  INIT VIEWS
    _vSave      = new ViewSave();
    _vCopy      = new ViewCopy();
    _vRender    = new ViewRender();
    _vSim       = new ViewSimulation(_treePos);
    
    Area viewport = gl::getViewport();
    gl::setMatrices(*_cameraOtho);
    gl::setViewport(_fboCurrent->getBounds());
    _fboCurrent->bindFramebuffer();
    _vSave->render();
    _fboCurrent->unbindFramebuffer();
    gl::setViewport(viewport);
    gl::setMatrices(*_camera);
    gl::rotate(_sceneQuat->quat);
    
}

void SubSceneParticles::update(){
    gl::disable(GL_DEPTH_TEST);
    gl::setMatrices(*_cameraOtho);
    Area viewport = gl::getViewport();
    gl::setViewport(_fboCurrent->getBounds());
    _fboTarget->bindFramebuffer();
    _vSim->render(_fboCurrent->getTexture());
    _fboTarget->unbindFramebuffer();
    gl::setViewport(viewport);
    gl::setMatrices(*_camera);
    gl::rotate(_sceneQuat->quat);
    gl::enable(GL_DEPTH_TEST);
    
//    writeImage("Fbo.png", _fboTarget->getTexture());
    
    swap(_fboTarget, _fboCurrent);
    _frameCount = 0;
}


void SubSceneParticles::render() {
    if(GlobalSettings::getInstance().showFbo) {
        gl::disable(GL_DEPTH_TEST);
        Area viewport = gl::getViewport();
        gl::setMatrices(*_cameraOtho);
        gl::setViewport(_fboCurrent->getBounds() + _fboCurrent->getBounds() + _fboCurrent->getBounds());
        _vCopy->render(_fboCurrent->getTexture());
        
        gl::enable(GL_DEPTH_TEST);
        gl::setViewport(viewport);
        gl::setMatrices(*_camera);
        gl::rotate(_sceneQuat->quat);
    }
    
    gl::enableAdditiveBlending();
    _vRender->render(_fboTarget->getTexture());
    gl::enableAlphaBlending();
}